runners/direct-java/src/main/java/org/apache/beam/runners/direct/DirectExecutionContext.java

```diff
@@ -81,6 +89,22 @@ public DirectTimerInternals timerInternals() {
       return timerInternals;
     }
 
+    @Override
+    public BundleFinalizer bundleFinalizer() {
+      if (bundleFinalizer == null) {
+        bundleFinalizer = new InMemoryBundleFinalizer();
```

Comment:
Why lazy-initialize? Is this just to save memory?

Reply:
yup since most transforms won't use the bundle finalizer

Reply:
Line#37 says the implementation is not thread-safe. So I assume there is no need to consider concurrent access to this initialization, right?